Add MCX-N5XX-EVK

decsny opened the pull request:
Add board target for the MCX-N5XX-EVK


To NXP board target enablers i.e. primarily @NeilChen93 @michal-smola @lucien-nxp @EmilioCBen @mmahadevan108 @butok @TomasGalbickaNXP @VitekST @ankeXiao @peterwangsz 
General note:  It is my opinion that when enabling boards that are very similar in design to an existing board, that they should be put in the same folder and share files instead of copy-pasting the same code and configurations into an entirely new folder.  Because doing the latter just causes there to have to be a lot more tedious refactoring later if something common between them has to change.  And they also each can't benefit from sharing these files, all of them end up becoming their own discrete projects that don't receive all the enablements together and it wastes a lot of time for everyone.  Especially boards that are showcasing derivative parts of other SOCs in the same series that there is already enablement of their EVKs, IMO they should seriously be put in the same folder to stop duplicating effort exponentially. Right now we especially have a huge explosion of `frdm_ke*`, `frdm_mcxa*`, and `frdm_mcxc*` folders that probably needs to stop and even should probably be reversed/combined in a way like I am doing on this PR.

The only complication I run into sometimes when doing this type of file sharing is mostly with the devicetree.  You can already see the specifics of how I did this one on the PR so I will not explain the details of why I organized the files this way and which ones to put certain nodes and whatnot, I think that is self explanatory.  But what can get unwieldy is if you have the wrong #include structure with the files.  So I will give a brief and simple overview of how to avoid this complication and make sure your DT files are defined completely solid. 

How to make the DT includes order solid and powerful to share code:
The DT uses C preprocessor to do includes, and the way files are included using this preprocessor must form a [directed acyclic](https://en.wikipedia.org/wiki/Directed_acyclic_graph) graph to avoid any compile error (obviously). I.e.,
Let
$G = (V, E)$
be the preprocessor include graph where 
$V = \{f_1, f_2, \dots, f_n\}$  
is the set of vertices (files, each $f_i$), and 
$E \subseteq V \times V$
is the set of directed edges, where an edge,
$(f_i, f_j) \in E$
means file $f_i$ includes file $f_j$ via a `#include` directive.

Then in order to avoid problems with confusing DT errors, all you need to do is make sure that $G$ is NOT [homeomorphic](https://en.wikipedia.org/wiki/Homeomorphism_%28graph_theory%29) with the [dipole graph](https://en.wikipedia.org/wiki/Dipole_graph), $D_m$, where $m$ is any arbitrary amount of edges in some [subgraph](https://en.wikipedia.org/wiki/Glossary_of_graph_theory#subgraph) $G' \cong \text{subdivision}(G)$ of $G$ that would be [isomorphic](https://en.wikipedia.org/wiki/Graph_isomorphism) with some $D_m$.

Example is if
$V(G') = \{f_1, f_2, f_3, f_4\}, \quad E(G') = \{(f_1, f_2), (f_2, f_4), (f_1, f_3), (f_3, f_4)\}$
this would cause some confusing problems, so just avoid that type of situation and everything should be rock solid.

The same logic also can and should be applied on the SOC level DTS as well, such as when I did a cleanup of the RT1170 includes tree (misnomer) which you can see here: https://github.com/zephyrproject-rtos/zephyr/commit/dba69f3f14e5b84b411062f689d781360cefd934

Theoretically header guards might avoid some of the same problems but I think we would want to have this type of structure anyways because it is the most coherent and powerful for maintainability and enabling convenience

TODO:
- [x] add board documentation for mcx_n5xx_evk. @DerekSnell request that each board makes sure to show up separately on the zephyr board list filter
- [x] requesting testing run by @hakehuang for mcx_n5xx_evk and mcx_n9xx_evk once the upstream CI passes

Commit:
boards: mcx_nx4x_evk: Add MCX-N5XX-EVK

Rename this board folder to a more generic name since some other similar
boards will be added.

Add MCX-N5XX-EVK to MCXN EVK board folder. It is an almost identical
board to MCXN 9XX EVK with some slightly different things populated and
of course having the MCXN547 SOC instead of MCXN947.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

hakehuang commented on boards/nxp/mcx_nxxx_evk/mcx_n5xx_evk-pinctrl.dtsi:
```diff
@@ -0,0 +1,7 @@
+ * Copyright 2025 NXP
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
missing pwm and lpcmp2 pin config

hakehuang commented on boards/nxp/mcx_nxxx_evk/mcx_n5xx_evk_mcxn547_cpu0.yaml:
```diff
@@ -0,0 +1,35 @@
+  - zephyr
+  - gnuarmemb
+supported:
+  - adc
```
I would not suggest to enable all features without checking the corresponding tests.

nordicjm commented on boards/nxp/mcx_nxxx_evk/Kconfig:
```diff
@@ -2,4 +2,9 @@
 # SPDX-License-Identifier: Apache-2.0
 
 config BOARD_MCX_N9XX_EVK
+	bool
+	select BOARD_EARLY_INIT_HOOK
+
+config BOARD_MCX_N5XX_EVK
+	bool
```
```suggestion
config BOARD_NEEDS_EARLY_HOOK
        bool
        default y if BOARD_MCX_N9XX_EVK || BOARD_MCX_N5XX_EVK
        select BOARD_EARLY_INIT_HOOK
```
or alternatively filter on e.g soc family, soc series, etc.

hakehuang commented:
@decsny kernel and hello world test PASS on n5xxx platform. CI failure is need rebase to latest code base

DerekSnell commented on boards/nxp/mcx_nxxx_evk/board.cmake:
```diff
@@ -27,6 +27,8 @@ elseif(CONFIG_SOC_MCXN947_CPU1)
   board_runner_args(linkserver  "--device=MCXN947:MCX-N9XX-EVK")
   board_runner_args(linkserver  "--core=cm33_core1")
+elseif(CONFIG_SOC_MCXN547_CPU0)
+  board_runner_args(linkserver  "--device=MCXN547:MCX-N5XX-EVK")
```
we need to add JLink support for the MCXN547 device

decsny replied:
okay, that can be done in a follow up PR

mmahadevan108 commented on boards/nxp/mcx_nxxx_evk/mcx_n5xx_evk_mcxn547_cpu0.yaml:
```diff
@@ -0,0 +1,16 @@
+flash: 2048
+toolchain:
+  - zephyr
+  - gnuarmemb
```
Are'nt there more features enabled in this board that should be added here?

decsny replied:
hake said to remove them in a comment above

kartben committed:
boards: nxp: mcx_n9xx_evk: deactivate eth phy on cpu1

deactivate eth phy on cpu1, because
mdio is also only activated on cpu0.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

ZhaoxiangJin commented on boards/nxp/mcx_nxxx_evk/board.c:
mcxn54x and mcxn94x can share the same board.c imo. mcxn94x is the master set, it has some more moto control related IP than mcxn54x, and the rest are the same.
BTW, the name of shared files should be MCXNx4x

Commit:
tests: hotfix mcxn qspi target exclude ram_context_for_isr

This platform is enabled incorrectly and it is not trivial to fix, but
need to hotfix CI because issue is related to linkage and causing a
build error

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

decsny commented:
@nordicjm can you please recheck this PR I tried to address all your comments, and now this PR is hotfixing 2 CI failure as well

decsny commented:
@hakehuang ci passes now

hakehuang reviewed (approved):
board testing pass

decsny commented:
@dleach02 @mmahadevan108 please review.
